### ISSUE TITLE
GIT_* env vars are not respected

### ISSUE THREAD

**Issue by svvac** [lambdalisue/gina.vim#130]
Gina should detect the `GIT_DIR` env var that points to a bare repository that is not located as per the default `/some/path/to/.git` scheme and use that if a better candidate was not found for the current file.

https://www.kernel.org/pub/software/scm/git/docs/#_the_git_repository

**Comment by lambdalisue** [comment]
Thanks. So you've meant

1. Find .git file/dir from current file (current behavior)
2. Find .git file/dir from real file if the current file is symlinked (current behavior)
3. Check GIT_DIR env and use it (new)

right? I'm wundering why GIT_DIR comes last. Shouldn't it come the first in the detection order?

**Comment by svvac** [comment]
Yeah, I thought about putting it first too, but since the env var is process-wide, editing a file in another repository would not setup the right one.

This is a bit of a brainstorming, but the following cases look sound to me:

Open vim with `GIT_DIR=~/bare/myrepo.git` and `CWD=~/workdir/myrepo`

```
~ (Filesystem boundary)
├── bare/
│   └── myrepo.git/     (bare git repo)
├── repos/
│   └── somedir/
│       ├── .git/       (bare git repo)
│       └── other_file.txt
├── unrelated/
│   └── dir/
│       └── file.txt
└── workdir/
    ├── elrepo/
    └── myrepo/
        └── some_file.txt
```

Case | Edited file                      | `GIT_DIR`             | `GIT_WORKDIR`         | **Result**
---- | -------------------------------- | --------------------- | -------------         | ----------
1.1  | `~/workdir/myrepo/some_file.txt` | -                     |-                      | No repository
1.2  | `~/workdir/myrepo/some_file.txt` | -                     | `~/workdir/myrepo`    | No repository
1.3  | `~/workdir/myrepo/some_file.txt` | `~/bare/myrepo.git`   | -                     | Use `GIT_DIR`
1.4  | `~/workdir/myrepo/some_file.txt` | `~/bare/myrepo.git`   | `~/workdir/myrepo`    | Use `GIT_DIR`
1.5  | `~/workdir/myrepo/some_file.txt` | `~/bare/myrepo.git`   | `~/workdir/elrepo`    | No repository
2.1  | `~/repos/somedir/other_file.txt` | -                     | -                     | Use `~/repos/somedir/.git`
2.2  | `~/repos/somedir/other_file.txt` | `~/bare/myrepo.git`   | -                     | Use `~/repos/somedir/.git`
2.4  | `~/repos/somedir/other_file.txt` | `~/bare/myrepo.git`   | `~/workdir/myrepo`    | Use `~/repos/somedir/.git`
2.5  | `~/repos/somedir/other_file.txt` | `~/bare/myrepo.git`   | `~/repos/somedir`     | Use `GIT_DIR`
3.1  | `~/unrelated/dir/file.txt`       | -                     | -                     | No repository
3.2  | `~/unrelated/dir/file.txt`       | `~/bare/myrepo.git`   | -                     | Use `GIT_DIR`
3.3  | `~/unrelated/dir/file.txt`       | `~/bare/myrepo.git`   | `~/workdir/myrepo`    | No repository

So basically:
  1. If `GIT_DIR` is set and `GIT_WORKDIR` is set and file under `GIT_WORKDIR` ⇒ use `GIT_DIR`;
  2. Climb up the file path looking for a `.git` folder. If found ⇒ use that.
  3. If file is a symlink, climb up the real file path looking for a `.git` folder. If found ⇒ use that.
  4. If `GIT_DIR` is set and `GIT_WORKDIR` is not set ⇒ use `GIT_DIR`
  5. No repository found.

**Comment by lambdalisue** [comment]
Ok. Please give me time to read & think :+1:

**Comment by lambdalisue** [comment]
Well. It's a bit too complicated and git detection runs a lot so I don't want to make it too complicated sorry.

p.s.

I said it is complicated because

- `GIT_DIR` or `GIT_WORK_TREE` or `GIT_COMMON_DIR` are required to be consider
- Above three variable requires to be an shortest absolute path (`simplify(fnamemodify(expand('xxx'), ':p'))`)
- How can I handle `GIT_COMMON_DIR`
- What's happen if `GIT_DIR` is correct but `GIT_WORK_TREE` is not
- etc.
